blog/_src/posts/2016-07-12-tutorial-zero-to-sixty-in-racket.scrbl

```diff
@@ -0,0 +1,543 @@
+@;  }
+@;]
+
+The first circular shift of a line @racket{A B C} is the line @racket{B C A}.
+The next circular shift is @racket{C A B}.
+
+Building a KWIK index is an important historical problem.
+According to @hyperlink["https://www.cs.umd.edu/class/spring2003/cmsc838p/Design/criteria.pdf"]{D.L. Parnas (1972)}:
+
+@nested{Except under extreme circumstances (huge data base, no supporting software)
+  such a system could be implemented by a good programmer within a week or two.
+}
+
+The year is 2016, and I bet only @hyperlink[TODOURL]{Agda} and @hyperlink[TODOURL]{PEX} programmers would
+ need the full two weeks.
+We'll be done in 20 minutes.
+
+
+@section{A Script}
+
+To start, open a file and type:
+
+@codeblock{
+  #lang racket
+}
+
+You can name the file anything, like @tt{kwik.rkt} or @tt{rkt.kwik} or @tt{foo}.
+Racket doesn't care, but it does need the @tt{#lang} line to compile the
+ contents of the file.
+
+The first part of the solution is a function to read input from a file into a
+ list of strings for further processing.
+The built-in function @racket[file->lines] does exactly this, but we'll
+ for-loop instead.
+
+@(begin
+  #reader scribble/comment-reader
+  (racketblock
+    (define (kwik-read filename)
+      (with-input-from-file filename
+        (λ ()
+          (for/list ([line (in-lines)])
+            line))))
+))
+
+@; The code is a little slanted, but Python has the same problem.
+
+When called with a filename like @|SAMPLE|, the function
+ uses @racket[for/list] to build a list of lines by reading from a port
+ with @racket[in-lines].
+The port is the data from @|SAMPLE| thanks to @racket[with-input-from-file].
+
+With a similar function, we can divide the strings into lists of words.
+Instead of writing this one by hand, we will just use library functions.
+
+@(begin
+  #reader scribble/comment-reader
+  (racketblock
+    (define (kwik-split lines)
+      (map string-split lines))
+))
+
+By default, @racket[string-split] divides a string into a list of whitespace-separated substrings.
+You can always supply a different delimiter, or use @racket[regexp-split]
+ to divide by a regular expression.
+
+There are two things left to do.
+First we generate all circular shifts for a list of strings @racket[words]
+ by folding up a list with one shift of @racket[words] for each word.
+
+@(begin
+  #reader scribble/comment-reader
+  (racketblock
+    (define (all-circular-shifts words)
+      (for/fold ([all-shifts (list words)])
+                ([i (in-range 1 (length words))])
+        (cons (circular-shift (first all-shifts))
+              all-shifts)))
+
+    (define (circular-shift words)
+      (append (rest words) (list (first words))))
+))
+
+Second, we alphabetize and print the shifts.
+
+@(begin
+ #reader scribble/comment-reader
+ (racketblock
+  (define (alphabetize all-shifts)
+    (sort all-shifts shift<?))
+
+  (define (shift<? shift1 shift2)
```

Comment:
Shouldn't (the version with the by-element comparison function as a parameter of) that be part of the standard library?

(Am I correct that this version assumes that `shift1` is always shorter than `shift2`?)